ci-operator/step-registry/cucushift/upgrade/tolatest/cucushift-upgrade-tolatest-commands.sh

```diff
@@ -0,0 +1,681 @@
+        if ! out="$(oc get clusterversion --no-headers)"; then continue; fi
+        avail="$(echo "${out}" | awk '{print $3}')"
+        progress="$(echo "${out}" | awk '{print $4}')"
+        if [[ ${avail} == "True" && ${progress} == "False" && ${out} == *"Cluster version is ${TARGET_VERSION}" ]]; then
+            echo -e "Upgrade succeed on $(date "+%F %T")\n\n"
+            return 0
+        fi
+    done
+    if [[ ${wait_upgrade} -le 0 ]]; then
+        echo -e "Upgrade timeout on $(date "+%F %T"), exiting\n" && return 1
+    fi
+}
+
+# Check version, state in history
+function check_history() {
+    local version state
+    version=$(oc get clusterversion/version -o jsonpath='{.status.history[0].version}')
+    state=$(oc get clusterversion/version -o jsonpath='{.status.history[0].state}')
+    if [[ ${version} == "${TARGET_VERSION}" && ${state} == "Completed" ]]; then
+        echo "History check PASSED, cluster is now upgraded to ${TARGET_VERSION}" && return 0
+    else
+        echo >&2 "History check FAILED, cluster upgrade to ${TARGET_VERSION} failed, current version is ${version}, exiting" && return 1
+    fi
+}
+
+if [[ -f "${SHARED_DIR}/kubeconfig" ]] ; then
+    export KUBECONFIG=${SHARED_DIR}/kubeconfig
+fi
+
+# Setup proxy if it's present in the shared dir
+if [[ -f "${SHARED_DIR}/proxy-conf.sh" ]]; then
+    # shellcheck disable=SC1091
+    source "${SHARED_DIR}/proxy-conf.sh"
+fi
+
+# Get the target upgrades release, by default, OPENSHIFT_UPGRADE_RELEASE_IMAGE_OVERRIDE is the target release
+# If it's serial upgrades then override-upgrade file will store the release and overrides OPENSHIFT_UPGRADE_RELEASE_IMAGE_OVERRIDE
+# upgrade-edge file expects a comma separated releases list like target_release1,target_release2,...
+export TARGET_RELEASES=("${OPENSHIFT_UPGRADE_RELEASE_IMAGE_OVERRIDE}")
+if [[ -f "${SHARED_DIR}/upgrade-edge" ]]; then
+    release_string="$(< "${SHARED_DIR}/upgrade-edge")"
+    # shellcheck disable=SC2207
+    TARGET_RELEASES=($(echo "$release_string" | tr ',' ' '))
+fi
+echo "Upgrade targets are ${TARGET_RELEASES[*]}"
+
+export OC="run_command_oc"
+
+# Target version oc will be extract in the /tmp/client directory, use it first
+mkdir -p /tmp/client
+export OC_DIR="/tmp/client"
+export PATH=${OC_DIR}:$PATH
+
+for target in "${TARGET_RELEASES[@]}"
+do
+    export TARGET="${target}"
+    # we need this DUMMY_TARGET_VERSION from ci config to download oc and do some pre-check
+    DUMMY_TARGET_VERSION="$(env "NO_PROXY=*" "no_proxy=*" oc adm release info "${TARGET}" --output=json | jq -r '.metadata.version')"
+    extract_oc
+
+    SOURCE_VERSION="$(oc get clusterversion --no-headers | awk '{print $2}')"
+    SOURCE_MINOR_VERSION="$(echo "${SOURCE_VERSION}" | cut -f2 -d.)"
+    export SOURCE_VERSION
+    export SOURCE_MINOR_VERSION
+    echo -e "Source release version is: ${SOURCE_VERSION}\n"
+
+    TARGET_MINOR_VERSION="$(echo "${DUMMY_TARGET_VERSION}" | cut -f2 -d.)"
+    export TARGET_MINOR_VERSION
```

Comment:
Just like what we discussed on slack, we might can determinate the exact `TARGET_VERSION` and `TARGET` before running `oc adm upgrade` command. The code is something like:
```
if ! check_ota_case_enabled "OCP-53907"; then
    set_channel
fi

SOURCE_VERSION="$(oc get clusterversion --no-headers | awk '{print $2}')"
SOURCE_MINOR_VERSION="$(echo "${SOURCE_VERSION}" | cut -f2 -d.)"
export SOURCE_VERSION
export SOURCE_MINOR_VERSION
echo -e "Source release version is: ${SOURCE_VERSION}\nSource minor version is: ${SOURCE_MINOR_VERSION}"

availableUpdates_file=$(mktemp)
echo "Available updates from CV:"
oc get clusterversion version -o json|jq -r '.status.availableUpdates' | tee "${availableUpdates_file}"
if [[ ! -s "${availableUpdates_file}" ]]; then
    echo "No recommended update available! exit..."
    exit 1
fi
availableUpdates_version_file=$(mktemp)
echo "Available update versions from CV:"
cat "${availableUpdates_file}" | jq -r "sort_by(.version) | reverse | .[].version" | tee "${availableUpdates_version_file}"
TARGET_VERSION=$(head -1 "${availableUpdates_version_file}")
if [[ -z "${TARGET_VERSION}" ]]; then
    echo "No recommended update version found! exit..."
    exit 1
fi
# for OCP version management, 4.x.y > 4.x.y-rc.n > 4.x.y-ec.n
# while sort only place the versions in alphabetical order
# so need this if block to get the correct target version
if [[ "$TARGET_VERSION" == *"-"* ]]; then
    temp_target_version="${TARGET_VERSION%-*}"
    if grep -q "^${temp_target_version}$" "${availableUpdates_version_file}"; then
        TARGET_VERSION="${temp_target_version}"
    fi
fi
TARGET_MINOR_VERSION="$(echo "${TARGET_VERSION}" | cut -f2 -d.)"
export TARGET_VERSION
export TARGET_MINOR_VERSION
echo -e "Target release version is: ${TARGET_VERSION}\nTarget minor version is: ${TARGET_MINOR_VERSION}"

target=$(cat "${availableUpdates_file}" | jq -r --arg z "${TARGET_VERSION}" '.[] | select(.version==$z) | .image')
export TARGET="${target}"
echo -e "Target release image is: ${TARGET}"
```

Reply:
After checked the latest codes, I saw change the codes a bit and move it in `set_target_and_upgrade_cmd` function. I want to explain my idea for the above codes.

Most of upgrade steps are developed based on `cucushift-upgrade-toimage` step, so here I almost copied the same code lines and layout from `cucushift-upgrade-toimage`, just for easily comparing code difference between steps in the future, in order to help locate potential script issues and code change sync. In the above code, the biggest change (compared with `cucushift-upgrade-toimage`) is to identify `TARGET_VERSION`, once the real `TARGET_VERSION` is confirmed, the real image also can be confirmed, save it in `TARGET`, then all other codes is almost the same as `cucushift-upgrade-toimage`. Once that, in the future, when comparing the difference, I just need to focus on the difference in the main part in this shell script.

Reply:
Hmm, i think here are two key points to change:
1. add the judgement of env ${UPGRADE_TO_VERSION}
2. determine the TARGET_VERSION
And the two points are cross-lap, so I add them into one func `set_target_and_upgrade_cmd` to make it more clear.